core/src/main/java/io/grpc/internal/OutboundState.java

```diff
@@ -0,0 +1,294 @@
+
+import io.grpc.Compressor;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.io.InputStream;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * A class that manages the outbound state on behalf of a stream. Manages the sending of headers,
+ * messages, and trailers, as well as outbound flow control.
+ */
+final class OutboundState {
+  /**
+   * The default number of queued bytes for a given stream, below which {@link
+   * StreamListener#onReady()} will be called.
+   */
+  private static final int ONREADY_THRESHOLD = 32 * 1024;
+
+  private enum Phase {
+    HEADERS, MESSAGE, STATUS
+  }
+
+  /**
+   * The handler for this state.
+   */
+  interface Handler {
+    /**
+     * Sends the given headers to the remote endpoint.
+     */
+    void sendHeaders(Metadata headers);
+
+    /**
+     * Sends the given message to the remote endpoint.
+     *
+     * @param frame the message to be sent.
+     * @param endOfStream if {@code true}, indicates that the frame should be marked as the end of stream.
+     * @param flush if {@code true}, indicates that any buffered IO should be flushed to the remote endpoint.
+     */
+    void sendMessage(@Nullable WritableBuffer frame, boolean endOfStream, boolean flush);
+
+    /**
+     * Sends the specified trailers to the remote endpoint. This is only called for server-side
+     * handlers.
+     *
+     * @param trailers the trailers to be sent
+     * @param headersSent if {@code true}, indicates that headers were previously sent.
+     */
+    void sendTrailers(Metadata trailers, boolean headersSent);
+
+    /**
+     * This indicates that the transport is now capable of sending additional messages
+     * without requiring excessive buffering internally. This event is
+     * just a suggestion and the application is free to ignore it, however doing so may
+     * result in excessive buffering within the transport.
+     */
+    void onReady();
+  }
+
+  private final boolean server;
+  private final Handler handler;
+  private final MessageFramer framer;
+  private Phase phase = Phase.HEADERS;
+  private boolean headersSent;
+  private Metadata stashedTrailers;
+
+  /**
+   * The number of bytes currently queued, waiting to be sent. When this falls below
+   * onReadyThreshold, {@link StreamListener#onReady()} will be called.
+   */
+  @GuardedBy("onReadyLock")
+  private int numSentBytesQueued;
+
+  private final Object onReadyLock = new Object();
+
+  OutboundState(WritableBufferAllocator bufferAllocator, Handler handler, boolean server) {
+    this.handler = checkNotNull(handler, "handler");
+    this.server = server;
+    MessageFramer.Sink frameHandler = new MessageFramer.Sink() {
+      @Override
+      public void deliverFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
+        onMessageFrame(frame, endOfStream, flush);
+      }
+    };
+    framer = new MessageFramer(frameHandler, bufferAllocator);
```

Comment:
Yep I share your pain.  I was originally calling it `OutboundContext`, but was worried it might cause confusion with the context API.  I'm definitely welcome to alternatives.